### DISCUSSION TITLE
Why enumeration types in generic constraints are not `unmanaged` and `struct`?

### DISCUSSION THREAD

**Discussion by kyoyama-kazusa** [dotnet/csharplang#3975] [Q&A] [answered]
Hello everyone. I found the documentation for unmanaged types. The documentation said that the enumeration types are also be unmanaged types. However, when I use generic constraint, I cannot use `sizeof(TEnum)` even if I wrote `where TEnum : Enum`. The compiler says, "Cannot take the address of, get the size of, or declare a pointer to a managed type ('TEnum')", which makes me confused.

If and only if I wrote the both constraint `Enum` and `unmanaged`, `sizeof(TEnum)` will work. But why? Is runtime problem?

```csharp
void F<TEnum>(TEnum a) where TEnum : unmanaged, Enum
{
    // ...
}
```

In addtion, `Enum`s cannot be reference types, but why we should add `struct` constraint to make some features work?

**Comment by HaloFour** [comment, marked as answer]
`System.Enum` is a weird beast in the runtime.  Even though it is the "base type" for all enums as far as the type system is concerned it is technically a reference type.  That's why you must combine it with the `struct` (or `unmanaged`) constraint which eliminates `System.Enum` itself as being used as the generic type argument.

**Reply by kyoyama-kazusa** [reply]
I got it. Thank you!
